File: ml-agents-envs/mlagents/envs/side_channel/side_channel.py
import Queue
from abc import ABC, abstractmethod

ChannelType = int


class SideChannel(ABC):
    """
    The side channel just get access to a bytes buffer that will be shared
    between C# and Python. For example, We will create a specific side channel
    for properties that will be a list of string (fixed size) to float number,
    that can be modified by both C# and Python. All side channels are passed
    to the Env object at construction.
    """

    def __init__(self):
        self.message_queue = Queue.Queue()

    def queue_message_to_send(self, data: bytearray) -> None:
        """
        Queues a message to be sent by the environment at the next call to
        step.
        """
        self.message_queue.put(data)

    @abstractmethod
    def on_message_received(self, data: bytearray) -> None:
        """
        Is called by the environment to the side channel. Can be called
        multiple times per step if multiple messages are meant for that
        SideChannel.
        """
        pass

    @property
    @abstractmethod
    def channel_type(self) -> ChannelType:
        """
        :return:The type of side channel used. Will influence how the data is
        processed in the environment.
        """
        pass

File: com.unity.ml-agents/Runtime/SideChannel/SideChannel.cs
using System.Collections.Generic;
using System;

namespace MLAgents
{
    public abstract class SideChannel
    {
        // The list of messages (byte arrays) that need to be sent to Python via the communicator.
        // Should only ever be read and cleared by a ICommunicator object.
        internal List<byte[]> MessageQueue = new List<byte[]>();

        /// <summary>
        /// An int identifier for the SideChannel. Ensures that there is only ever one side channel
        /// of each type. Ensure the Unity side channels will be linked to their Python equivalent.
        /// </summary>
        /// <returns> The integer identifier of the SideChannel</returns>
        public Guid ChannelId{
            get;
            protected set;
        }

        /// <summary>
        /// Is called by the communicator every time a message is received from Python by the SideChannel.
        /// Can be called multiple times per simulation step if multiple messages were sent.
        /// </summary>
        /// <param name="data"> the payload of the message.</param>
        public abstract void OnMessageReceived(byte[] data);

        /// <summary>
        /// Queues a message to be sent to Python during the next simulation step.
        /// </summary>
        /// <param name="data"> The byte array of data to be sent to Python.</param>
        protected void QueueMessageToSend(byte[] data)
        {
            MessageQueue.Add(data);
        }
    }
}
